tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.tla

```diff
@@ -0,0 +1,446 @@
+\* Copyright (c) 2024, Oracle and/or its affiliates.
+
+----------------------- MODULE BufferedRandomAccessFile -----------------------
+\* This is a model-checkable specification for BufferedRandomAccessFile.java.
+\* It covers the core fields as well as the seek, read, write, flush, and
+\* setLength operations.
+\*
+\* There are three major correctess conditions:
+\*
+\*   (1) the internal invariants V1-V5 should hold
+\*   (2) the behavior should refine a general RandomAccessFile
+\*   (3) each operation should refine its RandomAccessFile counterpart
+\*
+\* Readers will probably want to start with the general RandomAccessFile spec
+\* before reading this one.
+
+EXTENDS Naturals, Sequences, TLC, Common
+
+CONSTANT BuffSz
+
+VARIABLES
+    \* in-memory variables (BufferedRandomAccessFile class fields)
+    dirty,
+    length,
+    curr,
+    lo,
+    buff,
+    diskPos,
+
+    \* the underlying file
+    file_content,
+    file_pointer
+
+vars == <<
+    dirty, length, curr, lo, buff, diskPos,
+    file_content, file_pointer>>
+
+TypeOK ==
+    /\ dirty \in BOOLEAN
+    /\ length \in Offset
+    /\ curr \in Offset
+    /\ lo \in Offset
+    /\ buff \in Array(Symbols \union {ArbitrarySymbol}, BuffSz)
+    /\ diskPos \in Offset
+
+    /\ file_content \in ArrayOfAnyLength(Symbols \union {ArbitrarySymbol})
+    /\ ArrayLen(file_content) <= MaxOffset
+    /\ file_pointer \in Offset
+
+-------------------------------------------------------------------------------
+\* Internal invariants (copied from comment in BufferedRandomAccessFile.java)
+
+RelevantBufferContent ==
+    ArraySlice(buff, 0, Min(BuffSz, length - lo))
+
+LogicalFileContent == \* denoted c(f) in .java file
+    IF ArrayLen(RelevantBufferContent) > 0
+    THEN WriteToFile(file_content, lo, RelevantBufferContent)
+    ELSE file_content
+
+DiskF(i) == \* denoted disk(f)[i] in .java file
+    IF i >= 0 /\ i < ArrayLen(file_content)
+    THEN ArrayGet(file_content, i)
+    ELSE ArbitrarySymbol
+
+BufferedIndexes == lo .. (Min(lo + BuffSz, length) - 1)
+
+Inv1 ==
+    /\ diskPos = file_pointer
+    /\ length = ArrayLen(LogicalFileContent)
+
+\* Inv2 is a bit special.  Most methods restore it just before they return.  It
+\* is generally restored by calling `seeek(curr)`.  But, that behavior is
+\* difficult to model in straight TLA+ because each method may modify variables
+\* multiple times.  So instead, this spec treats Inv2 as a precondition for the
+\* methods and verifies that it is indeed restored by `Seek(pos)`, which should
```

Comment:
`Inv2CanAlwaysBeRestored` suggests that this should mention `FlushBuffer` in addition to `Seek`.